src/python/pants/engine/internals/build_files.py

```diff
@@ -387,16 +389,91 @@ def _extract_env_vars(
+            return tgt
+        return tgt.with_new_kwargs(**default_values, **tgt.kwargs)
+
+    name_to_path_and_synthetic_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for synthetic_address_map in synthetic_address_maps:
+        for name, target in synthetic_address_map.name_to_target_adaptor.items():
+            name_to_path_and_synthetic_target[name] = (
+                synthetic_address_map.path,
+                apply_defaults(target),
+            )
+
+    name_to_path_and_declared_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for declared_address_map in declared_address_maps:
+        for name, target in declared_address_map.name_to_target_adaptor.items():
+            name_to_path_and_declared_target[name] = (declared_address_map.path, target)
+
+    # We copy the dict so we can modify the original in the loop.
+    for name, (
+        declared_target_path,
+        declared_target,
+    ) in name_to_path_and_declared_target.copy().items():
+        declared_target_kwargs = dict(declared_target.kwargs)
+        extend_synthetic = declared_target_kwargs.pop("_extend_synthetic", False)
+        if extend_synthetic:
+            # Pop synthetic target to let the declared target take precedence.
+            synthetic_target_path, synthetic_target = name_to_path_and_synthetic_target.pop(
+                name, (None, None)
+            )
+            if synthetic_target is None:
+                raise InvalidTargetException(
+                    softwrap(
+                        f"""
+                            The `{declared_target.type_alias}` target {name!r} in {declared_target_path} has
+                            `_extend_synthetic=True` but there is no synthetic target to extend.
+                            """
+                    )
+                )
+
+            if synthetic_target.type_alias != declared_target.type_alias:
+                raise InvalidTargetException(
+                    softwrap(
+                        f"""
+                        The `{declared_target.type_alias}` target {name!r} in {declared_target_path} is
+                        of a different type than the synthetic target
+                        `{synthetic_target.type_alias}` from {synthetic_target_path}.
+
+                        When `_extend_synthetic` is true the target types must match, set this to
+                        false if you want to replace the synthetic target with the target from your
+                        BUILD file.
+                        """
+                    )
+                )
+
+            # Preserve synthetic field values not overriden by the declared target from the BUILD.
+            kwargs = dict(synthetic_target.kwargs)
+        else:
+            kwargs = {}
+        kwargs.update(declared_target_kwargs)
+        name_to_path_and_declared_target[name] = (
+            declared_target_path,
+            declared_target.with_new_kwargs(**kwargs),
+        )
```

Comment:
related to the previous perf question, this unconditionally creates a new target adaptor, even when there was no synthetic target at all. So this affects everything/everyone, not only targets using synthetic targets.

Reply:
Yep, fixed!